File: LuaLibs/src/MyLua/Libs/EventHandle/EventDispatchGroup.lua
require "MyLua.Libs.Core.GlobalNS"
require "MyLua.Libs.Core.Class"
require "MyLua.Libs.Core.GObject"
require "MyLua.Libs.DataStruct.Dictionary"

local M = GlobalNS.Class(GlobalNS.GObject:new())
M.clsName = "EventDispatchGroup"
GlobalNS[M.clsName] = M

function M:ctor()
    self.m_groupID2DispatchDic = GlobalNS.Dictionary:new()
    self.m_bInLoop = false;
end

-- 添加分发器
function M:addEventDispatch(groupID, disp)
    if not self.m_groupID2DispatchDic:ContainsKey(groupID) then
        self.m_groupID2DispatchDic:Add(groupID, disp);
    end
end

function M:addEventHandle(groupID, handle)
    self.m_groupID2DispatchDic:value(groupID):addEventHandle(handle)
end

function M:removeEventHandle(groupID, handle)
    if self.m_groupID2DispatchDic:ContainsKey(groupID) then
        self.m_groupID2DispatchDic:value(groupID):removeEventHandle(handle)
    else
        -- 日志
    end
end

function M:dispatchEvent(groupID,  dispatchObject)
    self.m_bInLoop = true;
    if self.m_groupID2DispatchDic:ContainsKey(groupID) then
        self.m_groupID2DispatchDic:value(groupID):dispatchEvent(dispatchObject)
    else
        -- 日志
    end
    self.m_bInLoop = false;
end

function M:clearAllEventHandle()
    if not m_bInLoop then
        for _, dispatch in pairs(self.m_groupID2DispatchDic) do
            dispatch:clearEventHandle()
        end

        self.m_groupID2DispatchDic:Clear()
    else
        -- 日志
    end
end

function M:clearGroupEventHandle(groupID)
    if not self.m_bInLoop then
        if self.m_groupID2DispatchDic:ContainsKey(groupID) then
            self.m_groupID2DispatchDic:value(groupID):clearEventHandle();
            self.m_groupID2DispatchDic:Remove(groupID);
        else
            -- 日志
        end
    else
        -- 日志
    end
end

return M

File: PyLibs/src/Libs/EventHandle/EventDispatchGroup.py
# -*- coding: utf-8 -*-

from Libs.Core.GObject import GObject;
from Libs.DataStruct.MDictionary import MDictionary;
from Libs.EventHandle.EventDispatch import EventDispatch;

class EventDispatchGroup(GObject):

    def __init__(self):
        super(EventDispatchGroup, self).__init__();
        
        self.mTypeId = "EventDispatchGroup";
        
        self.mGroupID2DispatchDic = MDictionary();
        self.mIsInLoop = False;          # 是否是在循环遍历中


    # 添加分发器
    def addEventDispatch(self, groupID, disp):
        if (not self.mGroupID2DispatchDic.ContainsKey(groupID)):
            self.mGroupID2DispatchDic.Add(groupID, disp);


    def addEventHandle(self, groupID, handle):
        # 如果没有就创建一个
        if (not self.mGroupID2DispatchDic.ContainsKey(groupID)):
            self.addEventDispatch(groupID, EventDispatch());

        self.mGroupID2DispatchDic.value(groupID).addEventHandle(None, handle);


    def removeEventHandle(self, groupID, handle):
        if (self.mGroupID2DispatchDic.ContainsKey(groupID)):
            self.mGroupID2DispatchDic.value(groupID).removeEventHandle(None, handle);

            # 如果已经没有了
            if (self.mGroupID2DispatchDic.value(groupID).getHandleCount() == 0):
                self.mGroupID2DispatchDic.Remove(groupID);
        else:
            # 输出日志
            pass;


    def dispatchEvent(self, groupID, dispatchObject):
        self.mIsInLoop = True;
        if (self.mGroupID2DispatchDic.ContainsKey(groupID)):
            self.mGroupID2DispatchDic.value(groupID).dispatchEvent(dispatchObject);
        else:
            # 输出日志
            pass;

        self.mIsInLoop = False;


    def clearAllEventHandle(self):
        if (not self.mIsInLoop):
            for (_, dispatch) in self.mGroupID2DispatchDic.items():
                dispatch.clearEventHandle();

            self.mGroupID2DispatchDic.Clear();
        else:
            # 输出日志
            pass;


    def clearGroupEventHandle(self, groupID):
        if (not self.mIsInLoop):
            if (self.mGroupID2DispatchDic.ContainsKey(groupID)):
                self.mGroupID2DispatchDic.value(groupID).clearEventHandle();
                self.mGroupID2DispatchDic.Remove(groupID);
            else:
                # 输出日志
                pass;
        else:
            # 输出日志
            pass;

